### ISSUE TITLE
DicomCleaner sav_dicom

### ISSUE THREAD

**Issue by sylvainKritter** [pydicom/deid#57]
Hello,
Thanks a lot for all these very usefull functions!
I get trouble generating dicom after clean:
  File "C:\deid\dicom\pixels\clean.py", line 181, in save_dicom
    dicom.PixelData = self.clean.tostring()
AttributeError: 'function' object has no attribute 'tostring'

Can you help me?
Regards

**Comment by vsoch** [comment]
Definitely! I think I know the issue. It's in the middle of the night where I am so I'll take a look tomorrow morning :-) Stay tuned!

**Comment by sylvainKritter** [comment]
Thanks a lot !

Good night !

Sylvain

 

De : Vanessa Sochat <notifications@github.com> 
Envoyé : mardi 12 juin 2018 09:57
À : pydicom/deid <deid@noreply.github.com>
Cc : Sylvain Kritter <sylvain@horasis.eu>; Author <author@noreply.github.com>
Objet : Re: [pydicom/deid] DicomCleaner sav_dicom (#57)

 

Definitely! I think I know the issue. It's in the middle of the night where I am so I'll take a look tomorrow morning :-) Stay tuned!

**Comment by vsoch** [comment]
hey @sylvainKritter here is a PR for you to test!

https://github.com/pydicom/deid/pull/58

You should be able to clone the branch and then install:

```bash
# Remove all previous versions
pip uninstall deid
# do the above command until it tells you it isn't installed

# Clone the repo to install the new version!
git clone -b fix/cleaned-bug https://github.com/pydicom/deid
cd deid
python setup.py install
```

When you confirm all is well, I'll merge and release on pypi. I'll also generate a DOI on zenodo in case you need to cite the software for your work.

**Comment by sylvainKritter** [comment]
Many thanks Vanessa,

This is working : it writes the dicom file.

Now I need to figure out why there is still names in the dicom file pixel area: I will re-read doc….

Best regards, thanks for being so quick to fix !!!

Sylvain

 

De : Vanessa Sochat <notifications@github.com> 
Envoyé : mardi 12 juin 2018 14:44
À : pydicom/deid <deid@noreply.github.com>
Cc : Sylvain Kritter <sylvain@horasis.eu>; Mention <mention@noreply.github.com>
Objet : Re: [pydicom/deid] DicomCleaner sav_dicom (#57)

 

hey @sylvainKritter <https://github.com/sylvainKritter>  here is a PR for you to test!

#58 <https://github.com/pydicom/deid/pull/58> 

You should be able to clone the branch and then install:

# Remove all previous versions
pip uninstall deid
# do the above command until it tells you it isn't installed
 
# Clone the repo to install the new version!
git clone -b fix/cleaned-bug https://github.com/pydicom/deid
cd deid
python setup.py install

When you confirm all is well, I'll merge and release on pypi. I'll also generate a DOI on zenodo in case you need to cite the software for your work.

**Comment by sylvainKritter** [comment]
Hi again Vanessa ,

Running your example deid/examples/dicom/pixels/run-cleaner-client.py

 

I got a clean png file, but for the dcm, I got error :

 

File "C:\pydicom\filewriter.py", line 425, in write_data_element

    raise ValueError('Pixel Data with undefined length must '

ValueError: Pixel Data with undefined length must start with an item tag

 

Regards,

sylvain

 

De : Vanessa Sochat <notifications@github.com> 
Envoyé : mardi 12 juin 2018 14:44
À : pydicom/deid <deid@noreply.github.com>
Cc : Sylvain Kritter <sylvain@horasis.eu>; Mention <mention@noreply.github.com>
Objet : Re: [pydicom/deid] DicomCleaner sav_dicom (#57)

 

hey @sylvainKritter <https://github.com/sylvainKritter>  here is a PR for you to test!

#58 <https://github.com/pydicom/deid/pull/58> 

You should be able to clone the branch and then install:

# Remove all previous versions
pip uninstall deid
# do the above command until it tells you it isn't installed
 
# Clone the repo to install the new version!
git clone -b fix/cleaned-bug https://github.com/pydicom/deid
cd deid
python setup.py install

When you confirm all is well, I'll merge and release on pypi. I'll also generate a DOI on zenodo in case you need to cite the software for your work.

**Comment by vsoch** [comment]
hey @sylvainKritter ! The algorithm to detect is based on the method used by CTP - there is a list of "known" locations in the deid recipe (see [here for the default](https://github.com/pydicom/deid/blob/master/deid/data/deid.dicom#L109)) and then they are blacked out with this function here: https://github.com/pydicom/deid/blob/master/deid/dicom/pixels/detect.py#L81 Matching to the recipe is based on header fields, so it follows that if your data isn't represented there, or the headers have been changed, it won't work well. A much better solution here would be to develop something that uses character recognition (and do away with the manual aspect), and this is a project I had wanted to do but I didn't have anyone to work with. I'll write the detail here in case someone reading (now or future) is interested in this work. We would want to:

 - generate a large set of labeled images (phi vs. not)
 - develop an OCR method (and actually I had started but didn't continue --> https://github.com/pydicom/dicom-scraper
 - test our method and compare to the manual, etc.

Ideally there would be a method that can be tested and then some actual numbers shown to various IRB boards so it could be approved. I had wanted to do this but I can only take on so much on my own! If you (or a group that you know) would be interested in this work let me know! For now I'll merge the PR, upload to pypi, and add the zenodo record. Thanks for finding the issue!
